prov/mrail/src/mrail.h

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright (c) 2018 Intel Corporation, Inc.  All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#if HAVE_CONFIG_H
+#  include <config.h>
+#endif /* HAVE_CONFIG_H */
+
+#include <string.h>
+
+#include <rdma/fabric.h>
+#include <rdma/fi_atomic.h>
+#include <rdma/fi_cm.h>
+#include <rdma/fi_domain.h>
+#include <rdma/fi_endpoint.h>
+#include <rdma/fi_eq.h>
+#include <rdma/fi_rma.h>
+#include <rdma/fi_tagged.h>
+#include "rdma/providers/fi_log.h"
+
+#include <ofi.h>
+#include <ofi_util.h>
+#include <ofi_proto.h>
+#include <ofi_prov.h>
+#include <ofi_enosys.h>
+
+#define MRAIL_MAJOR_VERSION 1
+#define MRAIL_MINOR_VERSION 0
+
+#define mrail_info_foreach(i, fi, info)					\
+	for (i = 0, fi = info->next;					\
+	     fi && strstr(fi->fabric_attr->prov_name, mrail_prov.name);	\
+	     fi = fi->next, i++)
+
+extern struct fi_info mrail_info;
+extern struct fi_provider mrail_prov;
+extern struct util_prov mrail_util_prov;
+extern struct fi_fabric_attr mrail_fabric_attr;
+
+struct mrail_fabric {
+	struct util_fabric util_fabric;
+	struct fid_fabric **rails;
+	size_t num_rails;
+};
+
+struct mrail_domain {
+	struct util_domain util_domain;
+	struct fid_domain **rails;
+	size_t num_rails;
```

Comment:
It's somewhat confusing to refer to fabrics and domains as 'rails'.  I think I understand the intent here, but maybe just saying 'fabrics / num_fabrics' and 'domains / num_domains' may be better.